File: kotlin/PlanesAndroid/app/src/main/java/com/planes/common/Coordinate2D.java
package com.planes.common;

import java.util.Objects;

public class Coordinate2D implements Cloneable {

    public Coordinate2D() {}
    public Coordinate2D(int x, int y) {
        m_x = x;
        m_y = y;
    }

    public int x() {
        return m_x;
    }

    public int y() {
        return m_y;
    }

    public Coordinate2D add(final Coordinate2D other) {
        return new Coordinate2D(m_x + other.m_x, m_y + other.m_y);
    }

    @Override
    public boolean equals(final Object other) {
        if (this == other)
            return true;
        if (other == null)
            return false;
        if (getClass() != other.getClass())
            return false;
        Coordinate2D coordinate = (Coordinate2D) other;

        return (m_x == coordinate.m_x) && (m_y == coordinate.m_y);
    }

    @Override
    public Object clone() {
        return new Coordinate2D(m_x, m_y);
    }

    private int m_x = 0;
    private int m_y = 0;
}

File: kotlin/PlanesCompose/app/src/main/java/com/planes/singleplayerengine/Coordinate2D.kt
package com.planes.singleplayerengine

class Coordinate2D(val x: Int, val y: Int) : Cloneable {
    private var m_x = 0
    private var m_y = 0

    init {
        m_x = x
        m_y = y
    }

    constructor(): this(0, 0)

    fun x(): Int {
        return m_x
    }

    fun y(): Int {
        return m_y
    }

    fun add(other: Coordinate2D): Coordinate2D {
        return Coordinate2D(m_x + other.m_x, m_y + other.m_y)
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other == null) return false
        if (javaClass != other.javaClass) return false
        val coordinate = other as Coordinate2D
        return m_x == coordinate.m_x && m_y == coordinate.m_y
    }

    public override fun clone(): Any {
        return Coordinate2D(m_x, m_y)
    }

    override fun hashCode(): Int {
        var result = m_x
        result = 31 * result + m_y
        return result
    }
}